### PR TITLE
OCPBUGS-35342: Add e2e test for connect timeout

### PR DESCRIPTION
**Pull request by alebedev87**
In order to be able to backport https://github.com/openshift/api/pull/1829, [I have been asked (by David Eads)](https://github.com/openshift/cluster-ingress-operator/pull/1084#issue-2347334722) to implement an e2e test. The ingress operator's e2e seems to be well suited for it similar to https://github.com/openshift/cluster-ingress-operator/pull/1013.

### CODE REVIEW

**Comment by alebedev87** [comment]
/test e2e-azure-operator


```
--- FAIL: TestAllowedSourceRangesStatus (204.22s)
FAIL
FAIL	github.com/openshift/cluster-ingress-operator/test/e2e	2268.228s
FAIL
make: *** [Makefile:56: test-e2e] Error 1
```

**Comment by alebedev87** [comment]
Accidentally clicked on "Close with comment".

**Comment by alebedev87** [comment]
/test e2e-azure-operator

```
--- FAIL: TestAllowedSourceRangesStatus (205.87s)
FAIL
FAIL	github.com/openshift/cluster-ingress-operator/test/e2e	2121.002s
FAIL
make: *** [Makefile:56: test-e2e] Error 1
```

File: test/e2e/operator_test.go [hunk 1]
```diff
@@ -3126,6 +3126,135 @@ func TestAWSELBConnectionIdleTimeout(t *testing.T) {
...
+		t.Fatal("failed to determine ingress operator deployment's image: ", err)
+	}
+
+	iptablesImage, err := getIptablesImage(t, kclient, 1*time.Minute)
+	if err != nil {
+		t.Fatal("failed to determine image with iptables tool: ", err)
+	}
+
+	// Create a pod with an HTTP application that delays the connection and sends echo responses.
+	httpdPod := buildDelayConnectHTTPPod("connect-timeout-http", operatorcontroller.DefaultOperandNamespace, iptablesImage, operatorImage)
```
Comment (gcs278) [inline test/e2e/operator_test.go:3156]: So I was scratching my head a bit. I forgot to update the updated Ingress Operator with the code here so it gets the `server-delay-connect` command, so the pod never starts and just crash loops.

But I had a successful run?
```
    operator_test.go:3247: got expected response code after unexpected elapsed time 219.607859ms
    operator_test.go:3247: got expected response code after unexpected elapsed time 200.351503ms
    operator_test.go:3251: got expected response code after elapsed time 8.201132105s
    operator_test.go:3256: deleted ingresscontroller test-connect-timeout
--- PASS: TestConnectTimeout (268.35s)
```

Must have been a freak false positive. Either way, how about ensuring the `connect-timeout-http` pod comes up and gets into the running state? At least, that will be helpful to debugging future E2E failure.

File: test/e2e/util_test.go [hunk 2]
```diff
@@ -244,6 +246,52 @@ func buildSlowHTTPDPod(name, namespace string) *corev1.Pod {
...
+		},
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Image: initImage,
+					Name:  "init-iptables",
+					Command: []string{
+						"/bin/sh",
+						"-c",
+						"iptables -I INPUT -p tcp --dport 8080 -m conntrack --ctstate NEW -j NFQUEUE --queue-num 100",
```
Comment (gcs278) [inline test/e2e/util_test.go:268]: Could you add a comment here explaining why this is necessary?

I think it's creating a NFQUEUE for the `serve-delay-connect-test-server` command to work correctly no?

File: test/http/serve-delay-connect.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/spf13/cobra"
+)
+
+const (
+	defaultPort     = "8080"
+	defaultDelay    = time.Duration(40 * time.Second)
+	defaultQueueNum = 100
+)
+
+func NewServeDelayConnectCommand() *cobra.Command {
```
Comment (gcs278) [inline test/http/serve-delay-connect.go:26]: I ran:
```
$ sudo ./ingress-operator serve-delay-connect-test-server
2024/06/12 16:32:46 Callback is registered with delay 40s
2024/06/12 16:32:46 Serving on 8080
```

Then I ran:
```
$ curl localhost:8080
HTTP/1.1
```

It didn't delay. Why is that? Do I need to run iptables commands for this function to work?

**Review by gcs278 — COMMENTED** [review]
Looks good to me, pretty much all of my comments are just nit picks. Feel free to take or leave them.

File: test/http/serve-delay-connect.go (hunk 3 above)
Comment (alebedev87) [reply test/http/serve-delay-connect.go:26]: Yes. A rule needs to send new tcp connections on `8080` port to the queue number used by the command (`100` by default). Let me add the exact command as a comment.

File: test/e2e/util_test.go (hunk 2 above)
Comment (alebedev87) [reply test/e2e/util_test.go:268]: Sure, will do.

File: test/e2e/util_test.go [hunk 4]
```diff
@@ -18,6 +18,8 @@ import (
 	operatorv1 "github.com/openshift/api/operator/v1"
 	routev1 "github.com/openshift/api/route/v1"
 
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/util_test.go:21]: Right, let me add this import to the `openshift` group.

Side note: I think openshift group should go last as it's less "generic". This reminds me of [this discussion](https://redhat-internal.slack.com/archives/CBWMXQJKD/p1678980874913529).

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -3126,6 +3126,135 @@ func TestAWSELBConnectionIdleTimeout(t *testing.T) {
...
+	if err := wait.PollImmediate(5*time.Second, 3*time.Minute, func() (bool, error) {
+		client := &http.Client{}
+		start := time.Now()
+		response, err := client.Do(request)
+		if err != nil {
+			t.Logf("got unexpected error while sending request: %v", err)
+			return false, nil
+		}
+		defer response.Body.Close()
+		// sC   The "timeout connect" stroke before a connection to the server could
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/operator_test.go:3230]: It's a copy paste from the reference link below. `sC` is the termination status code which we see in the haproxy access logs when the connection timeout kicks in. Here I tried to explain there reason I check for the codes 503, 504.

Let me rephrase it to be clearer.

File: test/e2e/operator_test.go (hunk 1 above)
Comment (alebedev87) [reply test/e2e/operator_test.go:3156]: That's no good. I tried to protect from the false positives by adding a range check for the elapsed time. Let me try to reproduce this and find a way to avoid.

File: test/http/serve-delay-connect.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	return command
+}
+
+func serveDelayConnect(queueNum uint16, delay time.Duration, port string) error {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (gcs278) [inline test/http/serve-delay-connect.go:57]: Hmm, I can't actually ctrl-c out of this program when I run it on the command line:
```
sudo ./ingress-operator serve-delay-connect-test-server
[sudo] password for gspence: 
2024/06/12 16:32:46 Callback is registered with delay 40s
2024/06/12 16:32:46 Serving on 8080
^[[1;5A^C2024/06/12 17:34:30 Received signal: interrupt
2024/06/12 17:34:30 Error reading from netlink socket: netlink receive: i/o timeout
^C^C^C^C
^C^C^C^C
```

I think it's not actually stopping the program. Does this do the same for you?

Comment (alebedev87) [reply test/http/serve-delay-connect.go:57]: Good catch! The problem was in the unnecessary signal handling, removed it.

File: test/e2e/operator_test.go (hunk 1 above)
Comment (alebedev87) [reply test/e2e/operator_test.go:3156]: `waitForPodReady` function added.

File: test/e2e/operator_test.go [hunk 7]
```diff
@@ -3126,6 +3126,135 @@ func TestAWSELBConnectionIdleTimeout(t *testing.T) {
...
+		t.Fatalf("failed to create route %s/%s: %v", route.Namespace, route.Name, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), route); err != nil {
+			t.Fatalf("failed to delete route %s/%s: %v", route.Namespace, route.Name, err)
+		}
+	}()
+
+	// Wait for the load balancer and DNS to be ready.
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, icName, availableConditionsForIngressControllerWithLoadBalancer...); err != nil {
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/operator_test.go]: Changed to 10 minutes.

File: test/e2e/operator_test.go [hunk 8]
```diff
@@ -3126,6 +3126,135 @@ func TestAWSELBConnectionIdleTimeout(t *testing.T) {
...
+	}
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	operatorImage, err := getIngressOperatorDeploymentImage(t, kclient, 1*time.Minute)
+	if err != nil {
+		t.Fatal("failed to determine ingress operator deployment's image: ", err)
+	}
+
+	iptablesImage, err := getIptablesImage(t, kclient, 1*time.Minute)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/operator_test.go:3151]: `getIptablesImage` now errors if the image was not found.

File: test/e2e/operator_test.go [hunk 9]
```diff
@@ -3126,6 +3126,135 @@ func TestAWSELBConnectionIdleTimeout(t *testing.T) {
...
+	ic.Spec.TuningOptions = operatorv1.IngressControllerTuningOptions{
+		ConnectTimeout: &metav1.Duration{Duration: 2 * time.Second},
+	}
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	operatorImage, err := getIngressOperatorDeploymentImage(t, kclient, 1*time.Minute)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/operator_test.go:3146]: `getIngressOperatorDeploymentImage` now errors if the image was not found.

File: test/e2e/operator_test.go [hunk 10]
```diff
@@ -3126,6 +3126,135 @@ func TestAWSELBConnectionIdleTimeout(t *testing.T) {
...
+	// instead of the default 5s to let the test run a little faster.
+	icName := types.NamespacedName{Namespace: operatorNamespace, Name: "test-connect-timeout"}
+	ic := newLoadBalancerController(icName, icName.Name+"."+dnsConfig.Spec.BaseDomain)
+	ic.Spec.TuningOptions = operatorv1.IngressControllerTuningOptions{
+		ConnectTimeout: &metav1.Duration{Duration: 2 * time.Second},
+	}
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+	defer assertIngressControllerDeleted(t, kclient, ic)
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/operator_test.go]: Done.

File: test/e2e/operator_test.go [hunk 11]
```diff
@@ -3126,6 +3126,139 @@ func TestAWSELBConnectionIdleTimeout(t *testing.T) {
...
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	// Get the ELB's hostname via the wildcard DNS record
+	wildcardRecordName := controller.WildcardDNSRecordName(ic)
+	wildcardRecord := &iov1.DNSRecord{}
+	if err := kclient.Get(context.Background(), wildcardRecordName, wildcardRecord); err != nil {
+		t.Fatalf("failed to get wildcard dnsrecord %s: %v", wildcardRecordName, err)
+	}
+	elbHostname := wildcardRecord.Spec.Targets[0]
```
Comment (gcs278) [inline test/e2e/operator_test.go:3202]: *nit* You imply ELBs here, but this test is not restricted to just AWS. Looks like it works fine on GCP and Azure, but shall we use a more generic term, like just lb or load balancer? I don't think they are called ELBs on GCP and azure.

Also, your comments mention ELBs as well.

File: test/http/serve-delay-connect.go (hunk 3 above)
Comment (gcs278) [reply test/http/serve-delay-connect.go:26]: Hmm, I ran it again, this time running:
```
$ sudo iptables -I INPUT -p tcp --dport 8080 -m conntrack --ctstate NEW -j NFQUEUE --queue-num 100
$ sudo ./ingress-operator serve-delay-connect-test-server
[sudo] password for gspence: 
2024/06/20 19:14:54 Callback is registered with delay 40s
2024/06/20 19:14:54 Serving on 8080
2024/06/20 19:14:56 Callback is triggered
2024/06/20 19:14:56 Delaying SYN packet for 40s
```
I curled in another terminal:
```
$ curl localhost:8080
HTTP/1.1
```
But immediately got an answer back. Seems like progress since I get log output for the `serve-delay-connect-test-server` command, but why would curl return immediately and not delay?

I suppose it could be some configuration with my laptop, because I did test with the route created by your E2E test, and it delays successfully. So I believe it works, but I was just curious why it wouldn't work on my local laptop.

EDIT: I figured it out. I need to use `curl 127.0.0.1:8080`, then it works. I wiresharked it and looks like curl used ipv6, which didn't work correctly. Not sure if it's worth a comment somewhere that IPv6 doesn't work correctly.

**Review by gcs278 — COMMENTED** [review]
Thanks for the responses. Just a few more nits, but looks nearly good to go.

File: test/http/serve-delay-connect.go (hunk 3 above)
Comment (alebedev87) [reply test/http/serve-delay-connect.go:26]: Right, it must be laptop specific because it works on my old RHEL. The golang's server is supposed to bind on ipv6 and ipv4 addresses for localhost, so the problem should be in the iptables. Would this work for you if the iptables command is run for ipv6?
```
sudo ip6tables -I INPUT -p tcp --dport 8080 -m conntrack --ctstate NEW -j NFQUEUE --queue-num 100
```

File: test/e2e/operator_test.go [hunk 12]
```diff
@@ -3979,6 +4112,30 @@ func waitForDeploymentCompleteWithOldPodTermination(t *testing.T, cl client.Clie
 	})
 }
 
+// waitForPodReady waits for a pod to become ready.
+func waitForPodReady(t *testing.T, cl client.Client, pod *corev1.Pod, timeout time.Duration) error {
+	t.Helper()
+	name := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+	p := &corev1.Pod{}
+	err := wait.PollUntilContextTimeout(context.TODO(), 2*time.Second, timeout, true, func(ctx context.Context) (bool, error) {
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/operator_test.go]: Done.

File: test/e2e/operator_test.go (hunk 11 above)
Comment (alebedev87) [reply test/e2e/operator_test.go]: Done.

File: test/e2e/operator_test.go [hunk 13]
```diff
@@ -3979,6 +4112,30 @@ func waitForDeploymentCompleteWithOldPodTermination(t *testing.T, cl client.Clie
 	})
 }
 
+// waitForPodReady waits for a pod to become ready.
+func waitForPodReady(t *testing.T, cl client.Client, pod *corev1.Pod, timeout time.Duration) error {
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply test/e2e/operator_test.go:4116]: I was thinking about but I didn't see any system in the way `waitFor*` functions are placed so I decided to keep it closer to other `waitFor` from the same file. May need a dedicated PR to reorder the util functions.

File: test/http/serve-delay-connect.go (hunk 3 above)
Comment (gcs278) [reply test/http/serve-delay-connect.go:26]: Ah yea that worked. Whoops, I forgot that iptables is ipv4 ipv6 specific.

File: test/e2e/util_test.go [hunk 14]
```diff
@@ -352,6 +433,21 @@ func getDeployment(t *testing.T, client client.Client, name types.NamespacedName
 	return &dep, nil
 }
 
+func getDaemonSet(t *testing.T, client client.Client, name types.NamespacedName, timeout time.Duration) (*appsv1.DaemonSet, error) {
+	t.Helper()
+	ds := appsv1.DaemonSet{}
+	if err := wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
```
Comment (gcs278) [inline test/e2e/util_test.go:439]: *nit* deprecated function.

**Review by gcs278 — COMMENTED** [review]
just one more nit, but LGTM. I'm fine to send it through either way.

File: test/e2e/util_test.go (hunk 14 above)
Comment (alebedev87) [reply test/e2e/util_test.go]: Done.

**Comment by gcs278** [comment]
Looks good!
/approve
/lgtm

**Comment by alebedev87** [comment]
Install failed:
```
INFO[2024-06-25T14:19:45Z] Step e2e-gcp-operator-ipi-install-install failed after 29s. 
```

/test e2e-gcp-operator

**Comment by alebedev87** [comment]
```
: [sig-network-edge][Conformance][Area:Networking][Feature:Router][apigroup:route.openshift.io] The HAProxy router should pass the h2spec conformance tests
```
This is little worrying. Let's see whether it's reproducible.

/test e2e-aws-ovn-single-node
